### PR TITLE
Fix some of the Sql functions + add tests for non-db part

### LINKED ISSUE #2605: Bug in Sql.Function Left?
Hi,

Just reading the code and found this:
```c#
[Sql.Function(                      PreferServerSide = true)]
[Sql.Function(PN.SQLite, "LeftStr", PreferServerSide = true)]
public static string? Left(string? str, int? length)
{
    return length == null || str == null || str.Length < length? null: str.Substring(1, length.Value);
}
```
Why startIndex = 1 in `str.Substring(1, length.Value)`?
Shouldn't be there startIndex = 0?
(startIndex = The zero-based starting character position of a substring in this instance.)

### PR DESCRIPTION
**Pull request by yarecky1**
Fixes #2605

### CODE REVIEW

File: Source/LinqToDB/Sql/Sql.cs [hunk 1]
```diff
@@ -508,7 +508,7 @@ public static bool Like(string? matchExpression, string? pattern, char? escapeCh
 		[Sql.Function(PN.SQLite, "LeftStr", PreferServerSide = true)]
 		public static string? Left(string? str, int? length)
 		{
-			return length == null || str == null || str.Length < length? null: str.Substring(1, length.Value);
+			return length == null || str == null || str.Length < length? null: str.Substring(0, length.Value);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Sql/Sql.cs:511]: I'm not sure that we should leave it such that `str.Length < length` means the return value is `null`. I think the result for that condition should be `str`. @MaceWindu please confirm?

Comment (MaceWindu) [reply Source/LinqToDB/Sql/Sql.cs:511]: Yes, if we mimic db behavior, we shold return str here

Comment (yarecky1) [reply Source/LinqToDB/Sql/Sql.cs:511]: I'll correct this, similar functions and tests as you propose.

**Comment by yarecky1** [comment]
@MaceWindu @viceroypenguin 
I will change
```c#
public static string? Substring(string? str, int? startIndex, int? length)
{
    return str == null || startIndex == null || length == null ?
           null :
	   str.Substring(startIndex.Value, length.Value);
}
```
to
```c#
public static string? Substring(string? str, int? startIndex, int? length)
{
    return str == null || startIndex == null || startIndex.Value < 1 || length == null ?
           null :
	   str.Substring(startIndex.Value - 1, length.Value);
}
```
Added: `startIndex.Value < 1` condition and `-1` to parameter `startIndex.Value`.

This is because database uses 1-based position parameter, not 0-based index.

Which also means, we should rename parameter from `startIndex` to just `start` to not mislead users. See [here](https://www.w3schools.com/sql/func_sqlserver_substring.asp). This will change 6 files in 24 places.

Do you agree?

**Comment by MaceWindu** [comment]
Agree

**Comment by viceroypenguin** [comment]
Agree

File: Source/LinqToDB/Sql/Sql.cs [hunk 2]
```diff
@@ -508,24 +521,35 @@ public static bool Like(string? matchExpression, string? pattern, char? escapeCh
...
 		public static string? Right(string? str, int? length)
 		{
-			return length == null || str == null || str.Length < length?
-				null :
-				str.Substring(str.Length - length.Value);
+			if (length == null || str == null) return null;
+			if (length.Value < 0)              return null;
+			if (length.Value > str.Length)     return str;
+
+			return str.Substring(str.Length - length.Value);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Sql/Sql.cs:539]:
```suggestion
			return str[^length.Value..];
```

File: Source/LinqToDB/Sql/Sql.cs [hunk 3]
```diff
@@ -508,24 +521,35 @@ public static bool Like(string? matchExpression, string? pattern, char? escapeCh
 		[Sql.Function(PN.SQLite, "LeftStr", PreferServerSide = true)]
 		public static string? Left(string? str, int? length)
 		{
-			return length == null || str == null || str.Length < length? null: str.Substring(1, length.Value);
+			if (length == null || str == null) return null;
+			if (length.Value < 0)              return null;
+			if (length.Value > str.Length)     return str;
+
+			return str.Substring(0, length.Value);
```
Comment (viceroypenguin) [inline Source/LinqToDB/Sql/Sql.cs:528]:
```suggestion
			return str[..length.Value];
```

File: Source/LinqToDB/Sql/Sql.cs (hunk 2 above)
Comment (yarecky1) [reply Source/LinqToDB/Sql/Sql.cs:539]: I'm not convinced of introducing new syntax but will not discuss :)

File: Source/LinqToDB/Sql/Sql.cs (hunk 3 above)
Comment (yarecky1) [reply Source/LinqToDB/Sql/Sql.cs:528]: I'm not convinced of introducing new syntax but will not discuss :)

File: Source/LinqToDB/Sql/Sql.cs (hunk 2 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Sql/Sql.cs:539]: Ah. Forgot that it requires extra types in earlier FX versions. :)

**Comment by yarecky1** [comment]
@MaceWindu OK. This is it. No more string functions to fix or worth to create tests.

File: Source/LinqToDB/Sql/Sql.cs [hunk 4]
```diff
@@ -415,9 +415,16 @@ public static TTo From<TFrom>(TFrom obj)
 		[Sql.Function  (PN.SQLite,   "Substr",                          PreferServerSide = true)]
 		[Sql.Expression(PN.Firebird, "Substring({0} from {1} for {2})", PreferServerSide = true)]
 		[Sql.Function  (PN.SapHana,  "Substring",                       PreferServerSide = true)]
-		public static string? Substring(string? str, int? startIndex, int? length)
+		public static string? Substring(string? str, int? start, int? length)
 		{
-			return str == null || startIndex == null || length == null ? null : str.Substring(startIndex.Value - 1, length.Value);
+			if (str == null || start == null || length == null) return null;
+			if (start.Value < 1 || start.Value > str.Length) return null;
```
Comment (MaceWindu) [inline Source/LinqToDB/Sql/Sql.cs:421]: that's is interesting case. e.g. various dbs behave differently. E.g. sqlserver doesn't return null for those cases it returns:
- empty string
- or part of string if zero or negative start used and start + length > 1

Let's accept this behavior for out-of-range parameters as it is not possible to emulate all dbs for it. Basically idea is: for valid parameters we implement expected behavior, for invalid parameters - behavior could differ from db behavior...

**Review by MaceWindu — APPROVED** [review]

File: Source/LinqToDB/Sql/Sql.cs (hunk 4 above)
Comment (yarecky1) [reply Source/LinqToDB/Sql/Sql.cs:421]: Ok, so let's keep it that way. One change here for parameter out of range will force me to change behaviour in other places. Some function doesn't exists in sqlserver so I had to check what happens then. I preferred to return null for all abnormal parameters unless I could confirm that function would return empty string.

**Comment by yarecky1** [comment]
Could somebody re-run azp?

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**